library/core/src/iter/adapters/zip.rs

```diff
@@ -200,6 +200,7 @@ where
         } else if A::may_have_side_effect() && self.index < self.a.size() {
             let i = self.index;
             self.index += 1;
+            self.len += 1;
```

Comment:
Doesn't this break the `TrustedLen` (and `TrustedRandomAccess`) impls? The "length" that is reported would not be constant any longer, and it would after this branch be bigger than the number of items being returned.

Or maybe I'm missing something :)

Reply:
Quite the opposite, the increment of `self.len` bilances the previous increment of `self.index`, keeping their difference (and `size_hint`) constant. 

The bug was in fact that incrementing `self.index` changed the `size_hint` when it shouldn't.

Reply:
Thanks! Might be worth adding a comment about that somewhere so the next person reading the code doesn't have the same question, and that it's ideally considered when changing something around that code in the future :)

Reply:
I was already considering writing a follow-up PR that properly documents the safety of ZipImpl, especially around the state of `index` and `len`

Reply:
Sounds like a good plan